### PR TITLE
feat: crc with automatic table generation

### PR DESCRIPTION
**Pull request by PatrickVKM**
Different polynomials can be used without generating the table manually

### CODE REVIEW

File: infra/util/Crc.hpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    template<typename CRC_TYPE, CRC_TYPE Polynomial, CRC_TYPE InitValue = 0, CRC_TYPE FinalXor = 0, bool ReflectInput = false, bool ReflectOutput = false>
+    class Crc
+    {
+        static_assert(ReflectInput == ReflectOutput, "Currently, non-matching ReflectInput and ReflectOutput is not supported");
+
+    public:
+        Crc() = default;
+        ~Crc() = default;
+
+        /// Update the CRC calculation with the next input byte.
```
Comment (richardapeters) [inline infra/util/Crc.hpp:43]: There are a lot of comments, but hardly any of them provide actual added value. Can you remove the ones that only use more words to say the same?

File: infra/util/Crc.hpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        void Update(uint8_t input)
```
Comment (richardapeters) [inline infra/util/Crc.hpp:44]: Can you separate the definition and declaration of methods, and move the definitions to the bottom of the file? That way, the interface of the class is more clearly on display.

**Comment by PatrickVKM** [comment]
Removed comments.
Moved implementation outside the class definition

**Review by richardapeters — APPROVED** [review]

File: infra/util/Crc.hpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        CRC_TYPE crc = InitValue;
+    };
+
+    using Crc8Maxim = Crc<uint8_t, 0x31, 0, 0, true, true>;
+    using Crc8Sensirion = Crc<uint8_t, 0x31, 0xff, 0, false, false>;
+    using Crc16Modbus = Crc<uint16_t, 0x8005, 0xffff, 0, true, true>;
+    using Crc16Xmodem = Crc<uint16_t, 0x1021, 0x0>;
+    using Crc16CcittFalse = Crc<uint16_t, 0x1021, 0xffff>;
+    using Crc64Ecma = Crc<uint64_t, 0x42F0E1EBA9EA3693, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, true, true>;
+
```
Comment (fabiangottstein) [inline infra/util/Crc.hpp:56]: Could you also add a ISO CRC32 typedef?
```cpp
    using Crc32 = Crc<uint32_t, 0x04C11DB7>;

```

File: infra/util/Crc.hpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    constexpr CRC_TYPE PolyReverse(CRC_TYPE Polynomial);
+
+    template<typename CRC_TYPE, CRC_TYPE Polynomial, CRC_TYPE InitValue = 0, CRC_TYPE FinalXor = 0, bool ReflectInput = false, bool ReflectOutput = false>
+    class Crc
+    {
+        static_assert(ReflectInput == ReflectOutput, "Currently, non-matching ReflectInput and ReflectOutput is not supported");
+
+    public:
+        Crc() = default;
+        ~Crc() = default;
```
Comment (fabiangottstein) [inline infra/util/Crc.hpp:21]: These `default`ed constructor and destructor can be omitted

File: infra/util/Crc.hpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            maskOut >>= 1;
+            maskIn <<= 1;
+        }
+        return ret;
+    }
+
+    template<typename CRC_TYPE, CRC_TYPE Polynomial, CRC_TYPE InitValue, CRC_TYPE FinalXor, bool ReflectInput, bool ReflectOutput>
+    void Crc<CRC_TYPE, Polynomial, InitValue, FinalXor, ReflectInput, ReflectOutput>::Update(uint8_t input)
+    {
+        if (ReflectInput)
```
Comment (fabiangottstein) [inline infra/util/Crc.hpp:76]: Use `if constexpr (..)` to ensure compile-time evaluation if possible. 
There are other if statements where this is applicable (e.g. `if constexpr (bitWidthEquals8)`)

File: infra/util/Crc.hpp (hunk 3 above)
Comment (PatrickVKM) [reply infra/util/Crc.hpp:54]: Used correct initvalue 0xffffffff. Added testcase

File: infra/util/Crc.hpp (hunk 4 above)
Comment (PatrickVKM) [reply infra/util/Crc.hpp]: removed

File: infra/util/Crc.hpp (hunk 5 above)
Comment (PatrickVKM) [reply infra/util/Crc.hpp]: removed

Comment (PatrickVKM) [reply infra/util/Crc.hpp]: Added constexpr for if-statements

File: infra/util/Crc.hpp (hunk 2 above)
Comment (PatrickVKM) [reply infra/util/Crc.hpp]: moved implementation out of the declaration

**Review by fabiangottstein — APPROVED** [review]

**Pull request merged**